use-cases.md

```diff
@@ -0,0 +1,184 @@
+---
+layout: page
+title: Instrumentation for Common Use Cases
+---
+<div id="toc"></div>
+
+This page aims to illustrate common use cases that developers who instrument their applications and libraries with OpenTracing API need to deal with.
+
+### Tracing a Function
+
+```python
+    def top_level_function():
+        span1 = tracer.start_trace('top_level_function')
+        try:
+            . . . # business logic
+        finally:
+            span1.finish()
+```
+
+As a follow-up, suppose that as part of the business logic above we call another `function2` that we also want to trace. In order to attach that function to the ongoing trace, we need a way to access `span1`. We discuss how it can be done later, for now let's assume we have a helper function `get_current_span` for that:
+
+```python
+    def function2():
+        span2 = get_current_span().start_child('function2') \
+            if get_current_span() else None
+        try:
+            . . . # business logic
+        finally:
+            if span2:
+                span2.finish()
+```
+
+We assume that for whatever reason develper does not want to start a new trace in this function if one hasn't been started by the caller already, so we account for `get_current_span` potentially returning `None`.
+
+These two examples are intentionally naive. Usually developers will not want to pollute their business functions directly with tracing code, but use other means like a [function decorator in Python](https://github.com/uber-common/opentracing-python-instrumentation/blob/master/opentracing_instrumentation/local_span.py#L59):
+
+```python
+    @traced_function
+    def top_level_function():
+        ... # business logic
+```
+
+### Tracing Server Endpoints
+
+When a server wants to trace execution of a request, it generally needs to go through these steps:
+
+  1. Attempt to deserialize Trace Context from the incoming request, in case the trace has already been started by the client.
+  1. Depending on the outcome, either start a new trace, or join the existing trace using deserialized Trace Context. This operation starts a new Span.
+  1. Store the newly created span in some _request context_ that is propagated throughout the application, either by application code, or by the RPC framework.
+  1. Finally, close the span using `span.finish()` when the server has finished processing the request.
+
+#### Deserializing Trace Conext from Incoming Request
+
+Let's assume that we have an HTTP server, and the Trace Context is propagated from the client via HTTP headers, accessible via `request.headers`:
+
+```python
+    context = tracer.trace_context_from_text(
+        trace_context_id=request.headers,
+        trace_attributes=request.headers
+    )
```

Comment:
The problem I have with this API design is really nothing more than the use of the words serialisation and deserialisation.

While serialisation is obviously a consequence to this behind the scenes, it's not actually what the user cares about here, and therefore it's not intuitive API.

What this is really about is the transforming the state from the protocol's object to the tracing's context.

And I would say that the serialisation is the writing of bytes or characters behind the scenes into streams/buffers.

Maybe we should be looking at how other frameworks and SPI handle such transformer objects.